### PR TITLE
NonceManager for ETH faucet

### PR DESCRIPTION
**Pull request by franckc**
### Description:

 - Slight refactor to use /eth rather than /eth.html as landing page
 - Implement NonceManager to deal with parallel transactions. Open to any better idea re how to handle this...

### Checklist:

- [X] Test your work and double-check to confirm that you didn't break anything
- [ ] Wrap any displayed ETH addresses with [`formattedAddress`](https://github.com/OriginProtocol/origin/blob/master/origin-dapp/src/utils/user.js#L15-L17)
- [ ] Wrap any new text/strings for translation
- [ ] Run `npm run translations` if there are any changes to translated strings
- [ ] Map any new environment variables with a default value in the Webpack config
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/origin-docs)

### CODE REVIEW

**Comment by franckc** [comment]
Actually this won't work since we have multiple instances of the faucet running. I'll either have use the DB as centralized storage or reduce number of instances of origin-faucet to 1... :(

**Comment by tomlinton** [comment]
I think reducing the number of faucet instances to 1 is OK. Two just provides zero downtime deployments (i.e. Kubernetes will take one down, update container and wait for it to start, then take the other down). I don't think this is that important for the faucet.

**Comment by tomlinton** [comment]
Could we not use a locking mechanism to make it wait between transactions? We have redis available.

**Comment by franckc** [comment]
We could lock using redis or the DB but then max throughput would be 1 txn per 15sec at most.
If we have 100 people in the audience, that means it would take 25min for all their txn to go thru.. :( probably not great.

**Comment by franckc** [comment]
And yes, agreed ok to reduce number of instances to 1 on origin-faucet.

**Comment by tomlinton** [comment]
True that, if you wanted to combine these ideas though we could store the nonce in redis, it can do atomic increments. Two faucets could then run side by side. Not as much hassle as the db but still a bit of a pain.

Might just be easier to have a single faucet. 😁

**Comment by franckc** [comment]
Improved NonceManager implementation to reduce chances it goes out of sync.

File: origin-faucet/src/eth.js [hunk 1]
```diff
@@ -103,42 +184,50 @@ class EthDistributor {
...
+      try {
+        // Calculate the nonce to use for this transaction.
+        const nonce = await this.nonceMgr.next()
+
+        // Issue the blockchain transaction.
+        logger.info(
+          `Blockchain call to send ${amount.toFixed()} to ${ethAddress}
+          from ${this.hotWalletAddress} with nonce ${nonce}`
+        )
+        const receipt = await await this.web3.eth.sendTransaction({
```
[the remark this answers is not in the document]
Comment (franckc) [reply origin-faucet/src/eth.js]: oh wow. how did this happen ! :) thanks

File: origin-faucet/src/eth.js [hunk 2]
```diff
@@ -1,11 +1,77 @@
...
 
 const logger = require('./logger')
 const db = require('./models')
 const enums = require('./enums')
 
+// HTML template for the /eth landing page.
+const html = fs.readFileSync(`${__dirname}/../public/eth.html`).toString()
+
+/**
+ * Singleton class for managing a monotonically increasing nonce
```
[the remark this answers is not in the document]
Comment (franckc) [reply origin-faucet/src/eth.js]: Yep, could use singleton. Given NonceManager is currently not a util but only used in this file, I think I'm going to keep it simple for now.

**Comment by franckc** [comment]
Good question @sparrowDom regarding what happens if a transaction in a batch fails. As far as I understand, if a transaction is submitted with a nonce too high, the Ethereum nodes buffers it until the hole in the nonce sequence is filled (see this [article](https://ethereum.stackexchange.com/questions/2808/what-happens-when-a-transaction-nonce-is-too-high))

I thought I had this covered but thinking more about it, the behavior is not good with the current implementation. Here is an example:
  - Batch 1 submitted with transactions having nonces N1, N2, N3, ... Nx
 - Transaction with nonce N3 fails
 - The Ethereum nodes buffers transactions N4...Nx waiting for nonce N3 to be filled
 - Since there are still pending transactions, the Nonce manager would generated nonces Nx+1, Nx+2, etc... for subsequent transactions submitted by the faucet.
 => the hole would not get filled and all those transactions would never get executed by the Ethereum nodes.

This is such a pain ! I wonder if it would be better to create a bunch of hot wallets and send transactions from there in a round robin way. The downside is that it would involve maintaining all those hot wallets and keep them funded which would be quite a pain from an operational standpoint... I guess it could be automated, like with a cronjob that refills the hot wallets periodically from a master hot wallet. Still that's extra complexity and operational risk...

**Comment by sparrowDom** [comment]
Perhaps an almost good solution would be to have a map of nonce values in in the NonceManager instead of a counter: 
```
const nonces = {
50: 'Pending',
51: 'Failed',
52: 'Completed',
53: 'Completed'
....
}
```

and instead of the highest nonce value the manager would return the highest value or a first Failed value. This way the "hole" would be filled. (Of course manager would need to update state when transaction is successful/failed)

Downside it doesn't work if users stop submitting transactions or faucet cap is reached and there is a hole. 

A bulletproof but complicated solution would be to fail all transactions after a failed transaction and resubmit them.

**Comment by franckc** [comment]
@sparrowDom I haven't merged yet :) Re-opening this PR...

**Comment by sparrowDom** [comment]
@franckc Ahh sorry my bad :) missclick

**Comment by franckc** [comment]
Yes, we could keep track of the nonce and associated transaction status... But regarding cancelling a transaction, it is not possible ! There is not way to withdraw a transaction once it has been received by an Ethereum node... :(

**Comment by sparrowDom** [comment]
hmm good point. I know Metamask allows you to resubmit transaction with higher gas. Perhaps resubmission also allows a change of nonce? Not sure about it...

**Comment by franckc** [comment]
No, the only way to cancel a transaction is to submit another transaction with the same nonce - and it is not even guarantee that will work since either transactions may finish...

I have added a safety mechanism to prevent too many pending transactions (which would be a symptom in case we have a nonce hole issue). Not perfect but at least this way we can detect failures and manually fix.

I'm going to land this for now to test the prod deployment and will keep working on a better solution in the background, which probably involves using your idea of keeping track of the nonces and associated transaction status and to use the the lowest nonce.

**Comment by sparrowDom** [comment]
Aha that's how resubmission of transactions work, cool

**Pull request merged**